docs/custom-transfers.md

```diff
@@ -0,0 +1,276 @@
+data to the process over stdin. This tells the process useful information about
+the configuration.
+
+The message will look like this:
+```json
+{ "event":"init", "operation":"download", "concurrent": true, "concurrenttransfers": 3 }
+```
+
+* `event`: Always "init" to identify this message
+* `operation`: will be "upload" or "download" depending on transfer direction
+* `concurrent`: reflects the value of `lfs.customtransfer.<name>.concurrent`, in
+  case the process needs to know
+* `concurrenttransfers`: reflects the value of `lfs.concurrenttransfers`, for if
+  the transfer process wants to implement its own concurrency and wants to
+  respect this setting.
+
+The transfer process should use the information it needs from the intiation
+structure, and also perform any one-off setup tasks it needs to do. It should
+then respond on stdout with a simple empty confirmation structure, as follows:
+
+```json
+{ }
+```
+
+Or if there was an error:
+
+```json
+{ "error": { "code": 32, "message": "Some init failure message" } }
+```
+
+#### Stage 2: 0..N Transfers
+
+After the initiation exchange, git-lfs will send any number of transfer 
+requests to the stdin of the transfer process. 
+
+##### Uploads
+
+For uploads the request sent from git-lfs to the transfer process will look
+like this:
+
+```json
+{ "event":"upload", "oid": "bf3e3e2af9366a3b704ae0c31de5afa64193ebabffde2091936ad2e7510bc03a", "size": 346232, "path": "/path/to/file.png", "action": { "href": "nfs://server/path", "header": { "key": "value" } } }
+```
+
+* `event`: Always "upload" to identify this message
+* `oid`: the identifier of the LFS object
+* `size`: the size of the LFS object
+* `path`: the file which the transfer process should read the upload data from
+* `action`: the "upload" action copied from the response from the batch API.
+  This contains "href" and "header" contents, which are named per HTTP
+  conventions, but can be interpreted however the custom transfer agent wishes
+  (this is an NFS example, but it doesn't even have to be an URL). Generally,
+  "href" will give the primary connection details, with "header" containing any
+  miscellaneous information needed.
+
+The transfer process should post one or more [progress messages](#progress) and 
+then a final completion message as follows:
+
+```json
+{ "event":"complete", "oid": "bf3e3e2af9366a3b704ae0c31de5afa64193ebabffde2091936ad2e7510bc03a"}
+```
+
+* `event`: Always "complete" to identify this message
+* `oid`: the identifier of the LFS object
+
+Or if there was an error in the transfer:
+
+```json
+{ "event":"complete", "oid": "bf3e3e2af9366a3b704ae0c31de5afa64193ebabffde2091936ad2e7510bc03a", "error": { "code": 2, "message": "Explain what happened to this transfer" }}
+```
+
+* `event`: Always "complete" to identify this message
+* `oid`: the identifier of the LFS object
+* `error`: Should contain a `code` and `message` explaining the error
```

Comment:
This documentation for `error` is slightly different than the documentation for `error` on the download complete message. This one doesn't say what error should be if there is no error, the other doesn't say what error should be if there is an error. And if there is no error _must_ it be null or can it also be absent?